framework/src/org/checkerframework/framework/type/BoundsInitializer.java

```diff
@@ -391,36 +454,76 @@ public void initializeExtendsBound(final AnnotatedWildcardType wildcard) {
-                    for (int i = 0; i < actualTypeArgs.size(); i++) {
-                        final AnnotatedTypeMirror annoTypeArg =
-                                AnnotatedTypeMirror.createType(
-                                        actualTypeArgs.get(i), declaredType.atypeFactory, false);
-
-                        final BoundPathNode node = addPathNode(new TypeArgNode(i));
-                        typeArgs.add(replaceOrVisit(annoTypeArg));
-                        removePathNode(node);
+                typeArgs = new ArrayList<>();
+                for (int i = 0; i < typeElement.getTypeParameters().size(); i++) {
+                    TypeMirror javaTypeArg =
+                            getJavaType(declaredType, typeElement.getTypeParameters(), i);
+                    final AnnotatedTypeMirror atmArg =
+                            AnnotatedTypeMirror.createType(
+                                    javaTypeArg, declaredType.atypeFactory, false);
+                    typeArgs.add(atmArg);
+                    if (atmArg.getKind() == TypeKind.WILDCARD && declaredType.wasRaw()) {
+                        ((AnnotatedWildcardType) atmArg).setUninferredTypeArgument();
                     }
                 }
-                declaredType.setTypeArguments(typeArgs);
             } else {
+                typeArgs = declaredType.typeArgs;
+            }
 
-                final List<AnnotatedTypeMirror> typeArgs =
-                        new ArrayList<>(declaredType.getTypeArguments());
-                final List<AnnotatedTypeMirror> typeArgReplacements =
-                        new ArrayList<>(typeArgs.size());
-                for (int i = 0; i < typeArgs.size(); i++) {
-                    final AnnotatedTypeMirror typeArg = typeArgs.get(i);
-                    final BoundPathNode node = addPathNode(new TypeArgNode(i));
-                    typeArgReplacements.add(replaceOrVisit(typeArg));
-                    removePathNode(node);
+            List<AnnotatedTypeMirror> typeArgReplacements = new ArrayList<>(typeArgs.size());
+            for (int i = 0; i < typeArgs.size(); i++) {
+                AnnotatedTypeMirror typeArg = typeArgs.get(i);
+                BoundPathNode node = addPathNode(new TypeArgNode(i));
+                if (typeArg.getKind() == TypeKind.WILDCARD) {
+                    ((AnnotatedWildcardType) typeArg)
+                            .setTypeVariable(typeElement.getTypeParameters().get(i));
                 }
+                typeArgReplacements.add(replaceOrVisit(typeArg));
+                removePathNode(node);
+            }
+
+            declaredType.setTypeArguments(typeArgReplacements);
+        }
 
-                declaredType.setTypeArguments(typeArgReplacements);
+        /**
+         * Store the wildcards created as type arguments to raw types. This way if the raw type is
+         * visited again, the same wildcard can be used.
```

Comment:
You said

> There's no need to limit or clear the cache as a new InitializerVisitor is created for each type that is initialized.

I'm now confused what the `visited again` means. Is it for a type like `Map<Raw, Raw>` where we have two raw types? Why should they get the same wildcard?
Or is the same InitializerVisitor used on the same type multiple times?

Reply:
```
class Foo<T extends Foo> {}
```

The upper bound of the wildcard in `Foo<?>` is `Foo`.   The type argument of `Foo` is initialized to `? extends Foo`.  The type argument of `Foo` in  `? extends Foo` needs to be initialized to the same type argument as the first `Foo` so that `BoundsInitializer.InitializerVisitor#replaceOrVisit` will return the cached AnnotatedWildcardType.

Reply:
How about adding this explanation somewhere, either here or with the class? Otherwise, the "visited again" isn't clear at all.